modules/services/mbsync.nix

```diff
@@ -75,35 +85,47 @@ in {
     };
   };
 
-  config = mkIf cfg.enable {
-    assertions = [
-      (lib.hm.assertions.assertPlatform "services.mbsync" pkgs
-        lib.platforms.linux)
-    ];
-
-    systemd.user.services.mbsync = {
-      Unit = { Description = "mbsync mailbox synchronization"; };
-
-      Service = {
-        Type = "oneshot";
-        ExecStart =
-          "${cfg.package}/bin/mbsync ${concatStringsSep " " mbsyncOptions}";
-      } // (optionalAttrs (cfg.postExec != null) {
-        ExecStartPost = cfg.postExec;
-      }) // (optionalAttrs (cfg.preExec != null) {
-        ExecStartPre = cfg.preExec;
-      });
-    };
+  config = mkIf cfg.enable (mkMerge [
+    (optionalAttrs (hasAttr "sytemd" options) {
+      systemd.user.services.mbsync = {
+        Unit = { Description = "mbsync mailbox synchronization"; };
+
+        Service = {
+          Type = "oneshot";
+          ExecStart =
+            "${cfg.package}/bin/mbsync ${concatStringsSep " " mbsyncOptions}";
+        } // (optionalAttrs (cfg.postExec != null) {
+          ExecStartPost = cfg.postExec;
+        }) // (optionalAttrs (cfg.preExec != null) {
+          ExecStartPre = cfg.preExec;
+        });
+      };
 
-    systemd.user.timers.mbsync = {
-      Unit = { Description = "mbsync mailbox synchronization"; };
+      systemd.user.timers.mbsync = {
+        Unit = { Description = "mbsync mailbox synchronization"; };
 
-      Timer = {
-        OnCalendar = cfg.frequency;
-        Unit = "mbsync.service";
-      };
+        Timer = {
+          OnCalendar = cfg.frequency;
+          Unit = "mbsync.service";
+        };
 
-      Install = { WantedBy = [ "timers.target" ]; };
-    };
-  };
+        Install = { WantedBy = [ "timers.target" ]; };
+      };
+    })
+    (optionalAttrs (hasAttr "launchd" options) {
+      launchd.agents.mbsync = {
+        enable = true;
+        config = ({
+          ProgramArguments = [ "${cfg.package}/bin/mbsync" ] ++ mbsyncOptions;
+          ProcessType = "Adaptive";
+          RunAtLoad = true;
+          StartInterval = 60
+            * 15; # TODO(ryantking): Use frequency here... somehow
```

Comment:
@ambroisie @rycee Any thoughts on how to handle the frequency? Launchd does support onCalendar, but you have to provide it in a dictionary, `<hour>00</hour><minute>00</minute>`, not just as a string like systems. I was playing around with writing a parser to parse from the string format to the dict for launchd, but that is questionable with the lackluster regex support that nix has, e.g., can't even do non-capturing groups.

What would you think about accepting a `or (str attrs)` as the type for frequency and instructing users which to use based on their system?

Reply:
Yeah, that is a bit of a conundrum. Once upon a time I made [a little library](https://github.com/NixOS/nixpkgs/pull/43852) to handle durations in Nix. So in principle it should be possible to handle, at least simple cases, in a generic way. It would be necessary to keep an escape hatch for people who'd like to use something like "Thu,Fri 2012-*-1,5 11:12:13". I will think about that and perhaps resurrect the library.

For now I imagine it would be simplest to update the frequency documentation and make the user enter `frequency = "<hour>00</hour><minute>00</minute>"`. Not very pretty but would probably work until we have something better.

Do you have any link to documentation about the interval expressions supported by launchd?